// nwb/lib/config/babel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.processBabelConfig = processBabelConfig;

var _babelPresetProposals = require("babel-preset-proposals");

var _chalk = _interopRequireDefault(require("chalk"));

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// TODO Remove - deprecated
let warnedAboutStageConfig = false;
const STAGE_3_PROPOSALS = {
  dynamicImport: true,
  importMeta: true,
  classProperties: true,
  numericSeparator: true,
  // Actually Stage 4, but still...
  exportNamespaceFrom: true
};
const STAGE_2_PROPOSALS = { ...STAGE_3_PROPOSALS,
  decorators: true,
  functionSent: true,
  logicalAssignmentOperators: true,
  throwExpressions: true
};
const STAGE_1_PROPOSALS = { ...STAGE_2_PROPOSALS,
  exportDefaultFrom: true,
  pipelineOperator: true,
  doExpressions: true
};
const PROPOSALS_BY_STAGE = {
  0: {
    all: true
  },
  1: STAGE_1_PROPOSALS,
  2: STAGE_2_PROPOSALS,
  3: STAGE_3_PROPOSALS
};

function processBabelConfig({
  report,
  userConfig
}) {
  let {
    cherryPick,
    env,
    loose,
    plugins,
    presets,
    proposals,
    removePropTypes,
    react,
    reactConstantElements,
    runtime,
    stage,
    config,
    ...unexpectedConfig
  } = userConfig.babel;
  let unexpectedProps = Object.keys(unexpectedConfig);

  if (unexpectedProps.length > 0) {
    report.error('babel', unexpectedProps.join(', '), `Unexpected prop${(0, _utils.pluralise)(unexpectedProps.length)} in ${_chalk.default.cyan('babel')} config - ` + 'see https://github.com/insin/nwb/blob/master/docs/Configuration.md#babel-configuration for supported config');
  } // cherryPick


  if ('cherryPick' in userConfig.babel) {
    if ((0, _utils.typeOf)(cherryPick) !== 'string' && (0, _utils.typeOf)(cherryPick) !== 'array') {
      report.error('babel.cherryPick', cherryPick, `Must be a ${_chalk.default.cyan('String')} or an ${_chalk.default.cyan('Array')}`);
    }
  } // env


  if ('env' in userConfig.babel) {
    if ((0, _utils.typeOf)(env) !== 'object') {
      report.error('babel.env', env, `Must be an ${_chalk.default.cyan('Object')}`);
    }
  } // loose


  if ('loose' in userConfig.babel) {
    if ((0, _utils.typeOf)(loose) !== 'boolean') {
      report.error('babel.loose', loose, `Must be ${_chalk.default.cyan('Boolean')}`);
    }
  } // plugins


  if ('plugins' in userConfig.babel) {
    if ((0, _utils.typeOf)(plugins) === 'string') {
      userConfig.babel.plugins = [plugins];
    } else if ((0, _utils.typeOf)(userConfig.babel.plugins) !== 'array') {
      report.error('babel.plugins', plugins, `Must be a ${_chalk.default.cyan('String')} or an ${_chalk.default.cyan('Array')}`);
    }
  } // presets


  if ('presets' in userConfig.babel) {
    if ((0, _utils.typeOf)(presets) === 'string') {
      userConfig.babel.presets = [presets];
    } else if ((0, _utils.typeOf)(presets) !== 'array') {
      report.error('babel.presets', presets, `Must be a ${_chalk.default.cyan('String')} or an ${_chalk.default.cyan('Array')}`);
    }
  } // proposals


  if ('proposals' in userConfig.babel) {
    if ((0, _utils.typeOf)(proposals) === 'object') {
      let errors = (0, _babelPresetProposals.validateOptions)(proposals);

      if (errors.length) {
        report.error('babel.proposals', null, errors.join('\n'));
      }
    } else if (proposals !== false) {
      report.error('babel.proposals', proposals, `Must be an ${_chalk.default.cyan('Object')} (to configure proposal plugins) or ` + `${_chalk.default.cyan('false')} (to disable use of proposal plugins)`);
    }
  } // removePropTypes


  if ('removePropTypes' in userConfig.babel) {
    if (removePropTypes !== false && (0, _utils.typeOf)(removePropTypes) !== 'object') {
      report.error(`babel.removePropTypes`, removePropTypes, `Must be ${_chalk.default.cyan('false')} (to disable removal of PropTypes) ` + `or an ${_chalk.default.cyan('Object')} (to configure react-remove-prop-types)`);
    }
  } // react


  if ('react' in userConfig.babel) {
    if ((0, _utils.typeOf)(react) === 'string') {
      userConfig.babel.react = {
        runtime: react
      };
    } else if ((0, _utils.typeOf)(react) !== 'object') {
      report.error(`babel.react`, react, `Must be an ${_chalk.default.cyan('Object')} (to configure @babel/preset-react options) ` + `or a ${_chalk.default.cyan('String')} (to configure @babel/preset-react's runtime option)`);
    }
  } // reactConstantElements


  if ('reactConstantElements' in userConfig.babel) {
    if ((0, _utils.typeOf)(reactConstantElements) !== 'boolean') {
      report.error('babel.reactConstantElements', reactConstantElements, `Must be ${_chalk.default.cyan('Boolean')}`);
    }
  } // runtime


  if ('runtime' in userConfig.babel) {
    if ((0, _utils.typeOf)(runtime) !== 'object' && runtime !== false) {
      report.error('babel.runtime', runtime, `Must be an ${_chalk.default.cyan('Object')} (to configure transform-runtime options) or ` + `${_chalk.default.cyan('false')} (to disable use of the runtime-transform plugin)`);
    }
  } // TODO Remove - deprecated
  // stage


  if ('stage' in userConfig.babel) {
    let hasProposalsConfig = ('proposals' in userConfig.babel);
    let deprecationMessages = ['Deprecated as of nwb v0.24.0, as Babel 7 no longer has preset-stage-X presets', 'Use babel.proposals config instead to enable the required proposal plugins'];

    if ((0, _utils.typeOf)(stage) === 'number') {
      if (stage < 0 || stage > 3) {
        report.error('babel.stage', stage, `Must be between ${_chalk.default.cyan(0)} and ${_chalk.default.cyan(3)}`);
      } else if (!hasProposalsConfig) {
        deprecationMessages.push(`For now, nwb will enable proposal plugins corresponding to the final version of preset-stage-${stage}`, `This is equivalent to the following babel.proposals config:`, '', ...(0, _utils.padLines)((0, _utils.toSource)(PROPOSALS_BY_STAGE[stage])).split('\n'));
        userConfig.babel.proposals = { ...PROPOSALS_BY_STAGE[stage]
        };
      }
    } else if (stage !== false) {
      report.error('babel.stage', stage, `Must be a ${_chalk.default.cyan('Number')} between ${_chalk.default.cyan(0)} and ${_chalk.default.cyan(3)} (to choose a stage preset), ` + `or ${_chalk.default.cyan('false')} (to disable use of a stage preset)`);
    } else {
      if (!hasProposalsConfig) {
        deprecationMessages.push(`For now, nwb will disable use of proposal plugins`, `This is equivalent to configuring ${_chalk.default.cyan('babel.proposals = false')}`);
        userConfig.babel.proposals = false;
      }
    }

    if (!warnedAboutStageConfig) {
      report.deprecated('babel.stage', ...deprecationMessages);
      warnedAboutStageConfig = true;
    }
  } // config


  if ('config' in userConfig.babel && (0, _utils.typeOf)(config) !== 'function') {
    report.error(`babel.config`, `type: ${(0, _utils.typeOf)(config)}`, `Must be a ${_chalk.default.cyan('Function')}`);
  }
}
